contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,464 @@
+	Enum                 []interface{}              `json:"enum,omitempty"`
+	MaxProperties        *int64                     `json:"maxProperties,omitempty"`
+	MinProperties        *int64                     `json:"minProperties,omitempty"`
+	Required             []string                   `json:"required,omitempty"`
+	Items                *JSONSchemaPropsOrArray    `json:"items,omitempty"`
+	AllOf                []JSONSchemaProps          `json:"allOf,omitempty"`
+	OneOf                []JSONSchemaProps          `json:"oneOf,omitempty"`
+	AnyOf                []JSONSchemaProps          `json:"anyOf,omitempty"`
+	Not                  *JSONSchemaProps           `json:"not,omitempty"`
+	Properties           map[string]JSONSchemaProps `json:"properties,omitempty"`
+	AdditionalProperties *JSONSchemaPropsOrBool     `json:"additionalProperties,omitempty"`
+	PatternProperties    map[string]JSONSchemaProps `json:"patternProperties,omitempty"`
+	Dependencies         Dependencies               `json:"dependencies,omitempty"`
+	AdditionalItems      *JSONSchemaPropsOrBool     `json:"additionalItems,omitempty"`
+	Definitions          Definitions                `json:"definitions,omitempty"`
+}
+
+// JSONSchemaRef represents a JSON reference that is potentially resolved.
+type JSONSchemaRef struct {
+	referencePointer Pointer
+	HasFullURL       bool
+	HasURLPathOnly   bool
+	HasFragmentOnly  bool
+	HasFileScheme    bool
+	HasFullFilePath  bool
+}
+
+// Pointer is the JSON pointer representation.
+type Pointer struct {
+	referenceTokens []string
+}
+
+// SchemaURL represents a schema url. Defaults to JSON Schema Specification Draft 4.
+type SchemaURL string
+
+const (
+	// JSONSchemaDraft4URL is the url for JSON Schema Specification Draft 4.
+	JSONSchemaDraft4URL SchemaURL = "http://json-schema.org/draft-04/schema#"
+)
+
+// StringOrArray represents a value that can either be a string or an array of strings.
+// Mainly here for serialization purposes.
+type StringOrArray []string
+
+// JSONSchemaPropsOrArray represents a value that can either be a JSONSchemaProps
+// or an array of JSONSchemaProps. Mainly here for serialization purposes.
+type JSONSchemaPropsOrArray struct {
+	Schema      *JSONSchemaProps
+	JSONSchemas []JSONSchemaProps
+}
+
+// JSONSchemaPropsOrBool represents JSONSchemaProps or a boolean value.
+// Defaults to true for the boolean property.
+type JSONSchemaPropsOrBool struct {
+	Allows bool
+	Schema *JSONSchemaProps
+}
+
+// Dependencies represent a dependencies property.
+type Dependencies map[string]JSONSchemaPropsOrStringArray
+
+// JSONSchemaPropsOrStringArray represents a JSONSchemaProps or a string array.
+type JSONSchemaPropsOrStringArray struct {
+	Schema   *JSONSchemaProps
+	Property []string
+}
+
+// Definitions contains the models explicitly defined in this spec.
+type Definitions map[string]JSONSchemaProps
+```
+
+Note: A reflective test to check for drift between the types and the conversion object for runtime usage will be added.
```

Comment:
.... between the types **here and the OpenAPI types** for runtime usage will be added.